CompilerSource/parsing/macros.cpp

```diff
@@ -0,0 +1,316 @@
+      size_t arg_num = 0;
+      if (found == params_by_name.end()) {
+        if (tokens[i].content == "__VA_OPT__") {
+          if (e != i) res.push_back(FuncComponent::TokenSpan{e, i});
+          res.push_back(FuncComponent::VAOpt{});
+          e = i + 1;
+        } else if (tokens[i].content == "__VA_ARGS__") {
+          arg_num = params.size() - 1;
+        } else {
+          continue;
+        }
+      } else {
+        arg_num = found->second;
+      }
+      if (e != i) res.push_back(FuncComponent::TokenSpan{e, i});
+      // Check if either neighboring token is CONCAT.
+      // We already know if the previous was CONCAT, because we'll have
+      // a PASTE at the end of our result vector. We must still look ahead.
+      size_t j = i;  // First, find next non-whitespace token.
+      while (++j < tokens.size() && tokens[j].PreprocessesAway());
+      // If the next meaningful token is CONCAT, or our last operation is
+      // PASTE, we use the raw argument token vector.
+      if ((j < tokens.size() && tokens[j].type == TTM_CONCAT) ||
+          (!res.empty() && res.back().tag == FuncComponent::PASTE)) {
+        res.push_back(FuncComponent::RawArgument{arg_num});
+      } else {
+        res.push_back(FuncComponent::ExpandedArgument{arg_num});
+      }
+      e = i + 1;
+    } else if (tokens[i].type == TTM_STRINGIFY) {
+      size_t j = i;
+      while (++j < tokens.size() && tokens[j].PreprocessesAway());
+      if (j >= tokens.size() || tokens[j].type != TT_VARNAME) {
+        herr->Error(tokens[i]) << "# must be followed by a parameter name";
+        continue;
+      }
+      if (e != i) res.push_back(FuncComponent::TokenSpan{e, i});
+      i = j;
+      auto found = params_by_name.find(tokens[i].content);
+      if (found == params_by_name.end()) {
+        herr->Error(tokens[i])
+            << "# must be followed by a parameter name; "
+            << tokens[i].content << " is not a parameter";
+        continue;
+      }
+      res.push_back(FuncComponent::Stringify{found->second});
+      e = i + 1;
+    }
+  }
+  if (e < tokens.size())
+    res.push_back(FuncComponent::TokenSpan{e, tokens.size()});
+  return res;
+}
+
+static string TokContentCat(const Token &left, const Token &right) {
+  string content;
+  content.reserve(left.content.size() + right.content.size());
+  content += left.content;
+  content += right.content;
+  return content;
+}
+
+static Token PasteTokens(
+    const Token &left, const Token &right, const ErrorContext &errc) {
+  static const MacroMap no_macros;
+  static const ParseContext empty_context(nullptr);
+  Lexer l(std::move(TokContentCat(left, right)), &empty_context, errc.herr);
+  Token res = l.ReadRawToken();
+  Token end = l.ReadRawToken();
+  if (end.type != TT_ENDOFCODE) {
+    errc.Error() << "Concatenation of `" << left.content
+                 << "` and `" << right.content
+                 << "` does not produce a valid token.";
+  }
+  return res;
+}
+/*
```

Comment:
Maybe you're fixing this soon but maybe you'll forget. IDK either way I think its a good practice to leave a comment for why its here in case you forget and we end up with a new copy later